experimental/iterators/test/lib/Conversion/OneToNTypeConversion/OneToNTypeConversion.cpp

```diff
@@ -0,0 +1,322 @@
+// needed for source, target, or argument materialization, so we do some
+// educated guessing to recover that information. Fixing both issues would
+// require to use a PatternRewriter that overloads various `notify*` functions
+// and similar and tracks all changes there. However, that would require a
+// dedicated pattern application driver, which is currently also left as an open
+// TODO.)
+LogicalResult applyOneToNConversion(Operation *op,
+                                    OneToNTypeConverter &typeConverter,
+                                    const FrozenRewritePatternSet &patterns) {
+  // Remember existing unrealized casts.
+  SmallSet<UnrealizedConversionCastOp, 4> existingCasts;
+  op->walk(
+      [&](UnrealizedConversionCastOp castOp) { existingCasts.insert(castOp); });
+
+  // Apply provided conversion patterns.
+  if (failed(applyPatternsAndFoldGreedily(op, patterns)))
+    return failure();
+
+  // Find all newly inserted unrealized casts (that haven't folded away).
+  SmallVector<UnrealizedConversionCastOp> worklist;
+  op->walk([&](UnrealizedConversionCastOp castOp) {
+    if (!existingCasts.contains(castOp))
+      worklist.push_back(castOp);
+  });
+
+  // Replace new casts with user materializations.
+  IRRewriter rewriter(op->getContext());
+  for (UnrealizedConversionCastOp castOp : worklist) {
+    // Create user materialization.
+    TypeRange resultTypes = castOp->getResultTypes();
+    rewriter.setInsertionPoint(castOp);
+    SmallVector<Value> materializedResults;
+
+    // Determine whether operands or results are already legal to know which
+    // kind of materilization this is.
+    ValueRange operands = castOp.getOperands();
+    bool areOperandTypesLegal = llvm::all_of(
+        operands.getTypes(), [&](Type t) { return typeConverter.isLegal(t); });
+    bool areResultsTypesLegal = llvm::all_of(
+        resultTypes, [&](Type t) { return typeConverter.isLegal(t); });
+
+    if (!areOperandTypesLegal && areResultsTypesLegal && operands.size() == 1) {
+      // This is a target materilization.
+      std::optional<SmallVector<Value>> maybeResults =
+          typeConverter.materializeTargetConversion(
+              rewriter, castOp->getLoc(), resultTypes, operands.front());
+      if (!maybeResults)
+        return failure();
+      materializedResults = maybeResults.value();
+    } else if (areOperandTypesLegal && !areResultsTypesLegal &&
+               resultTypes.size() == 1) {
+      // This is a source or an argument materialization.
+      std::optional<Value> maybeResult;
+      if (llvm::all_of(operands, [&](Value v) { return v.getDefiningOp(); })) {
+        // This is an source materialization.
+        maybeResult = typeConverter.materializeArgumentConversion(
+            rewriter, castOp->getLoc(), resultTypes.front(),
+            castOp.getOperands());
+      } else {
+        // This is an argument materialization.
```

Comment:
I'd have an assert here to catch the potential weird situation where operands are a mix of block arguments and op results because of, e.g., interplay of incorrect partial conversion foldings that folded away only a part of the IR.

Reply:
I am now unsure for which version of the code you made this comment but for *source* materializations, operands may actually be block arguments and there may be a mix of those and op results. I thus removed the cast again in the latest version.

If a cast is inserted due to a type conversion of an op and the operands of this op are (forwarded) block arguments, then that cast is a *source* conversion but its arguments are still block arguments. Example:

```mlir
func.func @materializations(%arg0: i1) -> tuple<i1> {
  %0 = "test.make_tuple"(%arg0) : (i1) -> tuple<i1>
  return %0 : tuple<i1>
}
```

The conversion pattern for `make_tuple` simply forwards its N operands as N results (without any op), so we get:

```mlir
func.func @materializations(%arg0: i1) -> tuple<i1> {
  %0 = builtin.unrealized_conversion_cast %arg0 : i1 to tuple<i1> {"kind" = "source"}
  return %0 : tuple<i1>
}
```